### DISCUSSION TITLE
KCP Automation - Programming Languages/ Frameworks to use

### DISCUSSION THREAD

**Discussion by kasturinarra** [kcp-dev/kcp#910] [Development]
We plan to automate basic sanity / smoke testing for KCP and  researching about programming languages/ frameworks to use. Current suggestions are  not to use gingko, bash and use built-in go test or may be python  ? Raising the discussion here to see if there are any other better suggestions/ideas about the programming language and frameworks which we can use,  Input would be a kubeconfig file we would just export it and run tests which are developed.

**Comment by ncdc** [comment]
@stevekuznetsov I'm sure you have some opinions here 😄

**Comment by ncdc** [comment]
Also: these tests are testing as we'd expect end users to interact with the system; namely, with `kubectl` and our plugins.

**Reply by xltian** [reply]
@ncdc  Hey, Andy, most other OpenShift components e2e testing are being automated with golang and ginkgo, any special reason that ginkgo framework won't be applicable for KCP?

**Reply by ncdc** [reply]
Hi @xltian, I don't know that I can speak for everyone, but at least some of us have found working with ginkgo to take more effort and to not to be as productive as the regular go testing package. If you can introduce patterns that make it easy to introduce new test and that don't make the ongoing maintenance burden too high, we potentially could consider ginkgo.

@kcp-dev/kcp-prototypers please share your thoughts if you have any. Thanks!

**Reply by marun** [reply]
I think use of ginkgo can be problematic from a maintainability perspective if the tests end up following the Behavior-Driven Development (BDD) pattern that ginkgo enables. This style of testing is suited to writing tests from an end-user perspective (e.g. `when I take this action, I expect this to happen`) but does not seem ideal for validating complex distributed systems like kube, openshift or kcp.

That said, I also don't think that go's built-in test framework is ideally suited to that kind of e2e testing either. It's well-suited to small-scale testing (e.g. unit), but there are a number of gaps that I think may become increasingly problematic for the project:

 - The process for a given test is isolated to the test package. That forces sharing state or context across a test run to be done external to the test process (e.g. prior to suite invocation or via the filesystem).
 - Similarly, each test package generates its own binary. afaink there's no easy way to compose a single binary that is composed of multiple packages - attempting to run `go test -c` against multiple packages will fail. This complicates the practice of shipping a known-good test binary that can be used by consumers (e.g. qa/customers/etc).

I think it may be worth considering switching the e2e suite to a tool that doesn't have the shortcomings listed above. Ginkgo is one such tool, and perhaps there are others. 

Note that a switch to ginkgo wouldn't require adoption of the BDD-style of testing. The kcp project could decide to adopt the parts of ginkgo that are productive (the suite composition and execution parts that address the shortcomings of `go test`) while retaining the relative simplicity and clarity of the current test suite.

**Comment by csams** [comment]
If the process is running some shell commands and interpreting the output, I second python for the job.

**Comment by kasturinarra** [comment]
Notes from yesterday's meeting: patterns that are used to use like gingko in bunch of the openshift qe tests may be we just say roll with it, i will make sure we get feedback and ideally we can make a decision today or tomorrow so that you are not held up any more. Thanks !!
cc: @ncdc

**Comment by ncdc** [comment]
@kasturinarra do you have an example of some ginkgo tests that are testing the CLI (`oc` and/or `kubectl`)?

**Reply by kasturinarra** [reply]
@ncdc here you go https://github.com/openshift/openshift-tests-private/blob/release-4.9/test/extended/workloads/cli.go, not sure if you can access, if cannot please let me know and i could copy paste that for you, thanks !!

**Reply by ncdc** [reply]
I can see it, thanks! @stevekuznetsov @sttts WDYT?

**Reply by stevekuznetsov** [reply]
That seems perfectly reasonable to me. Yes, Gingko is not ideal, but /shrug

**Reply by ncdc** [reply]
Ok, I would say proceed unless @sttts objects.

**Reply by sttts** [reply]
Am not opinionated about the tool so much. Am more interested in agreeing on where these tests will live. Do we plan github.com/kcp-dev/kcp-tests or similar?

**Reply by ncdc** [reply]
I'm fine with that. @kasturinarra is that ok with you & your team?

**Reply by xltian** [reply]
As long as  QE and Devel agree on contributing tire 1 automation testing to the same repo with the same automation framework/program lanaguage, then that should be fine for QE team.

**Reply by kasturinarra** [reply]
Yup that should be fine with us. Thanks !! So we would be going forward with gingko framework.
cc: @ncdc

**Reply by ncdc** [reply]
Do you want us to create a kcp-tests repo for you?

**Reply by kasturinarra** [reply]
can we hold it till monday if that is okay ?

**Reply by ncdc** [reply]
Sure

**Reply by kasturinarra** [reply]
@ncdc can we continue using https://github.com/kcp-dev/kcp/tree/main/test/e2e instead of a new repo ?

**Reply by ncdc** [reply]
@kasturinarra I don't know that we want to mix ginkgo and non-ginkto tests in the `test/e2e` packages, do we?

**Reply by kasturinarra** [reply]
@ncdc i am not sure if that would cause any friction, well if we do not want to we could create a new repo.

**Reply by ncdc** [reply]
If you're adding to `test/e2e`, I'd say your additions should be aligned with the patterns we're currently using. If you're looking to add different kinds of tests, we could create a separate directory for you (`test/functional` or `test/smoke` or `test/cli` or something like that), and you could use ginkgo in there. Or we can go with a separate repo. @sttts WDYT?

**Reply by kasturinarra** [reply]
@ncdc i think it should be okay to create a new repo, thanks !!

**Reply by ncdc** [reply]
https://github.com/kcp-dev/kcp-tests
